sklearn/utils/estimator_checks.py

```diff
@@ -2565,6 +2567,60 @@ def check_classifiers_regression_target(name, estimator_orig):
         assert_raises_regex(ValueError, msg, e.fit, X, y)
 
 
+@ignore_warnings(category=(DeprecationWarning, FutureWarning))
+def check_estimator_sparse_dense(name, estimator_orig):
+    rng = np.random.RandomState(52)
+    estimator = clone(estimator_orig)
+    estimator_sp = clone(estimator_orig)
+    tags = _safe_tags(estimator_orig)
+    centers = 2 if tags["binary_only"] else None
+    X, y = make_blobs(random_state=rng, cluster_std=0.5, centers=centers)
+    # for put some points to zero to have a little bit of sparsity
+    # TODO: update this with enforce_estimator_tags_X when #14705 is merged
+    if tags["requires_positive_X"]:
+        X -= np.min(X) - 1.
+    X_csr = sparse.csr_matrix(X)
+    if tags["multioutput"] or tags["multioutput_only"]:
+        y = y[:, np.newaxis]
+
+    for sparse_format in ['csr', 'csc', 'dok', 'lil', 'coo', 'dia', 'bsr']:
+        X_sp = X_csr.asformat(sparse_format)
+        if name in ['Scaler', 'StandardScaler']:
+            estimator.set_params(with_mean=False)
+            estimator_sp.set_params(with_mean=False)
+        if name in ['RobustScaler']:
+            estimator.set_params(with_centering=False)
+            estimator_sp.set_params(with_centering=False)
+
+        set_random_state(estimator)
+        set_random_state(estimator_sp)
+        X_converted = pairwise_estimator_convert_X(X, estimator)
+        X_sp_converted = pairwise_estimator_convert_X(X_sp, estimator)
+        try:
+            estimator_sp.fit(X_sp_converted, y)
+            estimator.fit(X_converted, y)
+            if hasattr(estimator, "predict"):
+                pred = estimator.predict(X_converted)
+                pred_sp = estimator_sp.predict(X_sp_converted)
+                if not (isinstance(estimator, BaseSGD) or  # SGD is too random
```

Comment:
I wish we didn't need to make these exceptions by name. But it's not clear which exceptions deserve having tags named for them. Now that we have decomposed check_estimator, I would like to try implementing these exceptions in test_common rather than here... But maybe that belongs in another pr

Reply:
I am not really familiar with `test_common` but it seems to be the place to run tests with particular values for estimators parameters, is that it ? So do you mean that we should run a test for equivalence between sparse and dense for `StandardScaler` in `test_common` rather than in `sklearn.preprocessing.tests.test_data` for instance ?

Reply:
Estimator checks are used both internally in scikit-learn to ensure that all estimators have a consistent behaviors and error messages for edge cases. But they are also used by third party libraries to check that their estimators are compatible with the scikit-learn way of doing things. Because those checks can be used by third party libraries, we would rather like them not depend on specific estimators in scikit-learn. In that respect the `isinstance(estimator, BaseSGD)` is a red flag.

If some estimators should skip a particular test for good reason, we usually tag them with a dedicated estimator tag:

https://scikit-learn.org/stable/developers/develop.html#estimator-tags

See in particular `_xfail_checks`.

The scikit-learn internal thing that run the estimator checks on scikit-learn estimators is the `test_common`, see: `sklearn/tests/test_common.py`.

Reply:
Still I find it intriguing that the randomness of SGD has an impact on this check. We call `set_random_state(estimator)` and `set_random_state(estimator_sp)` so this is not a good reason in itself. Maybe there is a real bug with this estimator.